develop-py-plugins.md

```diff
@@ -1,7 +1,22 @@
 ## How to develop Python plugins
 
-We are currently writing up this section, will be updated soon.
+Python plugins use the [JSON-RPC](https://flow-launcher.github.io/docs/#/json-rpc) protocol to communicate with Flow via JSON structured calls.
 
-For now you can use the [Python template](https://github.com/Flow-Launcher/Flow.Launcher.Plugin.PythonTemplate) to set up your plugin 
+When building a Python plugins there are several things to be mindful of:
 
-Follow this Python plugin as an [example](https://github.com/deefrawley/Flow.Launcher.Plugin.Currency) 
+* The most important thing is we do not expect users to have to manually install the dependencies in requirements.txt because we aim to provide a seamless experience for them. This can be achieved by adding the following three things to your project:
+    1. Add a GitHub workflow- use a GitHub workflow that will install all your plugin's depedencies including the Python flowlauncher module to a folder called Lib inside your plugin.
+    2. Publish all as a zip- zip up your project including a lib directory that contains the modules and publish it to GitHub Releases page.
+    3. Point your module imports to the lib directory- reference all the modules to that directory where they are first imported.
+
+* The user can use their own system installed Python with Flow Launcher. But in most circumstances the user will most likely be using Flow Launcher's own embedded Python executable. [Embedded Python](https://docs.python.org/3/using/windows.html#the-embeddable-package) is isolated from the users system and does not prepend the scripts run directory to the system `PATH`.<sup>[ref](https://bugs.python.org/issue28245)</sup> If you need to import external files please follow the example below.
+
+### Simple Example
+Have a look at this simple example plugin [here](https://github.com/Flow-Launcher/plugin-samples/tree/master/HelloWorldPython), notice it has a folder called '.github/workflows' and a file called 'Publish Release.yml'. This is the workflow file that GitHub Workflow uses to run the CICD for the project. Moving out of that folder you can go into the [main.py](https://github.com/Flow-Launcher/plugin-samples/blob/master/HelloWorldPython/plugin.json) file, this is the entry file for your plugin. Notice it has this code block:
+```python
+import sys,os
+parent_folder_path = os.path.abspath(os.path.dirname(__file__))
+sys.path.append(parent_folder_path)
+sys.path.append(os.path.join(parent_folder_path, 'lib'))
```

Comment:
```python
sys.path.append(parent_folder_path)
sys.path.append(os.path.join(parent_folder_path, 'plugin'))
```
Should be delete.
Inserting lib folder path is fine, other things are useless.

Reply:
Another reason is that I thought the following one should stay in a zip file, not in the project.
To separate development and product two status.
The following is the product status, not the development.
These are a kind of mixing them up.

```python
import os
import sys

basedir = os.path.dirname(os.path.abspath(__file__))
sys.path.append(os.path.join(basedir, "lib"))


from plugin import Main

if __name__ == "__main__":
    Main()
```

Reply:
I think we can safely delete this line `sys.path.append(os.path.join(parent_folder_path, 'plugin'))`

But this line `sys.path.append(parent_folder_path)` have to stay because `sys.path` does not contain it, if you delete it you will see error like this:
```
  File "C:\Users\user1\AppData\Roaming\FlowLauncher\Plugins\Search-MDI-2.1.1\main.py", line 13, in <module>
    from plugin.main import MDI
ModuleNotFoundError: No module named 'plugin'
```

> I thought the following one should stay in a zip file, not in the project.
> To separate development and product two status.
> The following is the product status, not the development.
> These are a kind of mixing them up.

About this, I still prefer we keep it simple, avoid CI "magic" or new commands in this, I understand the idea about separate dev and prod environment, but Flow is a small personal tool not a big enterprise system, we should try to avoid unnecessary abstractions.

Abstractions come with costs :smile: Sometimes repeat code is better than try to abstract them to a command or classes or functions, because good abstractions is hard! 😄

Reply:
Unfortunately you need to add both to the `sys.path` to compensate for embedded Python.

Reply:
i think the plugin entry is not needed, with the currency plugin, i removed it, and was still able to use it.

Reply:
The interpreter to run a script what it needs to know is the **path**.

I don't think it belongs to `sys.path` problem.
It should be the **plugin folder** and the **entry script** relative position that causes this problem.

Is there a link to this plugin?

>   File "C:\Users\user1\AppData\Roaming\FlowLauncher\Plugins\Search-MDI-2.1.1\main.py", line 13, in <module>
>     from plugin.main import MDI
> ModuleNotFoundError: No module named 'plugin'

Reply:
Search-MDI is my plugin, codes [here](https://github.com/Garulf/Search-MDI). I haven't updated it to work with embedded Python yet.

Reply:
> The interpreter to run a script what it needs to know is the **path**.
> 
> I don't think it belongs to `sys.path` problem.
> It should be the **plugin folder** and the **entry script** relative position that causes this problem.
> 
> Is there a link to this plugin?
> 
> > File "C:\Users\user1\AppData\Roaming\FlowLauncher\Plugins\Search-MDI-2.1.1\main.py", line 13, in 
> > from plugin.main import MDI
> > ModuleNotFoundError: No module named 'plugin'

I think it's related to how EmbeddablePython is designed, but I'm not sure, just skimming these:

https://bugs.python.org/issue34841
https://stackoverflow.com/a/52566528/4230390
https://stackoverflow.com/questions/44443254/python-embeddable-zip-file-doesnt-include-lib-site-packages-in-sys-path

@Flow-Launcher/team @Garulf If you guys interested in remove 3 boilerplate lines mutate `sys.path`, probably have to read these links, I think with a bit of change in `python38._pth` file, we can remove the 3 lines, but we gotta be sure what are the implications of that :smile:

Reply:
> I think it's related to how EmbeddablePython is designed, but I'm not sure, just skimming these:

Oh it's 100% how embedded Python isolates itself from the users system. I've been fooling around with `_pth` files in hopes of removing the need for unnecessary boiler plate. So far I havent gotten it to work.

Reply:
> thanks @pc223 !

No problem!

> You still need to add lib to the sys.path for imports, which is very reasonable.

Agreed, I too think it's reasonable to retain that line 👍 Seems really explicit about how Flow deal with Python external dependencies.

> python38.zip
> .
> 
> Uncomment to run site.main() automatically
> import site

> Can this be done on installation of the embedded package by Flow?

This can be related to my bundled `pip` proposal :smile: In my CI script below, you can see I add `import site` to the `python38._pth` file to be able to install `pip` with `get-pip.py` 😄 
https://github.com/pc223/PythonEmbeddableWithPip/blob/main/.github/workflows/test2.yml
<img width=350 src="https://user-images.githubusercontent.com/10551242/129306619-807dc41e-f990-47ee-b95d-1ab01291e07c.png">

Reply:
@Garulf Hm...On my machine I have to **delete** the `python38._pth` to make it to work

Reply:
>@Garulf Hm...On my machine I have to delete the python38._pth to make it to work

`FlowLauncher/app-1.8.1/UserData/PythonEmbeddable/python38._pth`?

Reply:
yes

Reply:
Odd, if anything I'd assume it to _not_ work when absent.